File: WebClient SDK/src/default/cards/openStageSettingsOnDoubleClickExtension.ts
import { extension } from '@tessa/application';
import { Guid } from '@tessa/core';
import { showLoadingOverlay } from 'tessa/ui';
import { GridViewModel } from 'tessa/ui/cards/controls/grid/gridViewModel';
import { openCard } from 'tessa/ui/uiHost';
import { DoubleClickInfo, openCardDoubleClickAction } from 'tessa/ui/views';
import { WorkplaceViewComponentExtension } from 'tessa/ui/views/extensions';
import { IWorkplaceViewComponent } from 'tessa/ui/views/workplaceViewComponent';

/**
 * Расширение, открывающее параметры этапа по двойному клику по строке представления <b>KrStageRows</b>.
 */
@extension()
export class OpenStageSettingsOnDoubleClickExtension extends WorkplaceViewComponentExtension {
  //#region base overrides

  override getExtensionName(): string {
    return 'Tessa.Extensions.Default.Client.Cards.OpenStageSettingsOnDoubleClickExtension';
  }

  override initialize(model: IWorkplaceViewComponent): void {
    if (!model.doubleClickAction || model.inSelectionMode()) {
      return;
    }

    model.doubleClickAction = async (info: DoubleClickInfo) => {
      await openCardDoubleClickAction(info, async (cardId, displayValue, context) => {
        const selectedRow = info.context.viewContext?.selectedRow;
        if (!selectedRow) {
          return;
        }

        const stageRowID = selectedRow.get('StageRowID');
        if (!Guid.isValid(stageRowID)) {
          return;
        }

        const editor = await showLoadingOverlay(async () => {
          return await openCard({
            cardId,
            displayValue,
            context
          });
        });

        if (!editor) {
          return;
        }

        const cardModel = editor.cardModel!;

        const approvalProcessTab = cardModel.mainFormWithTabs?.tabs.find(
          i => i.name == 'ApprovalProcess'
        );
        if (!approvalProcessTab) {
          return;
        }

        const grid = cardModel.controls.get('ApprovalStagesTable') as GridViewModel;
        if (!grid) {
          return;
        }

        const stageRowViewModel = grid.rows.find(i => Guid.equals(i.rowId, stageRowID));
        if (!stageRowViewModel) {
          return;
        }

        cardModel.mainFormWithTabs!.selectedTab = approvalProcessTab;

        if (grid.selectedRow && grid.selectedRow !== stageRowViewModel) {
          grid.selectedRow.isSelected = false;
        }
        stageRowViewModel.isSelected = true;

        await grid.editRow(stageRowViewModel);
      });
    };
  }

  //#endregion
}

File: Extensions/Tessa.Extensions.Default.Client/Cards/OpenStageSettingsOnDoubleClickExtension.cs
#nullable enable

using System;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Runtime;
using Tessa.UI;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Controls;
using Tessa.UI.Views;

namespace Tessa.Extensions.Default.Client.Cards
{
    /// <summary>
    /// Расширение, открывающее параметры этапа по двойному клику по строке представления <b>KrStageRows</b>.
    /// </summary>
    /// <remarks>
    /// У расширения есть конфигуратор <see cref="OpenStageSettingsOnDoubleClickExtensionConfigurator"/>
    /// </remarks>
    public sealed class OpenStageSettingsOnDoubleClickExtension :
        IWorkplaceViewComponentExtension
    {
        #region Fields

        private readonly ISession session;
        private readonly IUIHost uiHost;

        #endregion

        #region Constructors

        public OpenStageSettingsOnDoubleClickExtension(
            ISession session,
            IUIHost uiHost)
        {
            this.session = NotNullOrThrow(session);
            this.uiHost = NotNullOrThrow(uiHost);
        }

        #endregion

        #region IWorkplaceViewComponentExtension Members

        /// <inheritdoc/>
        public void Initialize(
            IWorkplaceViewComponent model)
        {
            if (this.session.Token?.ApplicationID == ApplicationIdentifiers.TessaAdmin
                || model.InSelectionMode())
            {
                return;
            }

            model.DoubleClickAction = new DoubleClickAction(this.uiHost);
        }

        /// <inheritdoc/>
        public void Initialized(
            IWorkplaceViewComponent model)
        {
        }

        /// <inheritdoc/>
        public void Clone(
            IWorkplaceViewComponent source,
            IWorkplaceViewComponent cloned,
            ICloneableContext context)
        {
        }

        #endregion

        #region DoubleClickAction Private Class

        private sealed class DoubleClickAction :
            OpenCardDoubleClickAction
        {
            #region Fields

            private readonly IUIHost uiHost;

            #endregion

            #region Constructors

            public DoubleClickAction(IUIHost uiHost) =>
                this.uiHost = uiHost;

            #endregion

            #region Base Overrides

            /// <inheritdoc/>
            protected override async Task OpenCardAsync(
                Guid cardID,
                string? displayValue,
                IUIContext context,
                ViewDoubleClickInfo info,
                Guid? cardTypeID = null,
                string? cardTypeName = null)
            {
                var selectedRow = context.ViewContext.SelectedRow;
                if (selectedRow is null
                    || !selectedRow.TryGetValue("StageRowID", out var stageRowIDObj)
                    || stageRowIDObj is not Guid stageRowID)
                {
                    return;
                }

                ICardUIContextObject? cardUIContextObject;
                using (var splash = TessaSplash.Create(TessaSplashMessage.OpeningCard))
                {
                    cardUIContextObject = await this.uiHost.OpenCardAsync(
                        cardID,
                        options: new OpenCardOptions
                        {
                            DisplayValue = displayValue,
                            UIContext = context,
                            Splash = splash,
                        });
                }

                if (cardUIContextObject is null)
                {
                    return;
                }

                var cardModel = cardUIContextObject.Context.CardEditor.CardModel;

                var approvalProcessTab = cardModel
                    .MainFormWithTabs
                    .Tabs
                    .FirstOrDefault(i => i.Name == KrConstants.Ui.KrApprovalProcessFormAlias);

                if (approvalProcessTab is null
                    || !cardModel.Controls.TryGet(KrConstants.Ui.KrApprovalStagesControlAlias, out var control)
                    || control is not GridViewModel grid
                    || grid.Rows.FirstOrDefault(i => i.Model.RowID == stageRowID) is not { } stageRowViewModel)
                {
                    return;
                }

                cardModel.MainFormWithTabs.SelectedTab = approvalProcessTab;
                grid.SelectedRow = stageRowViewModel;

                await grid.EditRowAsync(stageRowViewModel);
            }

            #endregion
        }

        #endregion
    }
}
